# Aplicando funciones a un arreglo.

¿Cómo puedes calcular transformaciones y estadísticas sobre todos los elementos de un arreglo de forma rápida?

Una de las mayores fortalezas de NumPy es la posibilidad de aplicar funciones matemáticas y estadísticas irectamente sobre arreglos, sin tener que escribir bucles. Esto permite procesar grandes volúmenes de datos con muy pocas líneas de código y con un rendimiento muy superior al de las listas tradicionales de Python.
Estas funciones están optimizadas en C y se ejecutan de forma vectorizada, es decir, se aplican automáticamente a todos los elementos del arreglo.

## Funciones matemáticas elementales

NumPy incluye funciones básicas como:

In [9]:
import numpy as np

In [10]:
datos = np.array([1, 4, 9, 16])

print(np.sqrt(datos)) # Raíz cuadrada: [1. 2. 3. 4.]
print(np.log(datos)) # Logaritmo natural
print(np.exp(datos)) # Exponencial e^x
print(np.sin(datos)) # Seno de cada elemento (en radianes no grados)
print(np.cos(datos)) # Coseno de cada elemento (en radianes no grados)

[1. 2. 3. 4.]
[0.         1.38629436 2.19722458 2.77258872]
[2.71828183e+00 5.45981500e+01 8.10308393e+03 8.88611052e+06]
[ 0.84147098 -0.7568025   0.41211849 -0.28790332]
[ 0.54030231 -0.65364362 -0.91113026 -0.95765948]


Cada una de estas funciones transforma todos los elementos del arreglo de forma simultánea y sin bucles.

## Aplicar funciones estadísticas

Estas funciones son esenciales para el análisis exploratorio de datos.

In [11]:
valores = np.array([10, 20, 30, 40])

print(np.mean(valores)) # Promedio: 25.0
print(np.std(valores)) # Desviación estándar: 11.18
print(np.min(valores)) # Valor mínimo: 10
print(np.max(valores)) # Valor máximo: 40
print(np.sum(valores)) # Suma total: 100

25.0
11.180339887498949
10
40
100


la desviación estandar es el promedio de las difrencias de cada dato con el preomedio.

In [12]:
print(10-25)
print(20-25)
print(30-25)
print(40-25)

-15
-5
5
15


In [13]:
# saco el cuadrado de lo anterio  
print(-15*-15)
print(-5*-5)
print(5*5)
print(15*15)

225
25
25
225


In [14]:
#luego sumo lo anterior 
print(225+25+25+225)

500


In [15]:
# luego saco el promedio de la suma 
print(500/4)

125.0


In [16]:
# luego saco la raiz cuadrada
print(np.sqrt(125))

11.180339887498949


## Aplicar funciones sobre matrices

Cuando se trabaja con matrices (2D), NumPy permite aplicar funciones a lo largo de un eje:

In [17]:
matriz = np.array([[1, 2, 3],[4, 5, 6]])

print(matriz)

print(np.sum(matriz, axis=0))# Suma por columna: [1+4= 5  2+5= 7  3+6= 9 ] ---> [5 7 9]
print(np.sum(matriz, axis=1))# Suma por fila: [1+2+3 = 6  4+5+6 = 15 ]---> [6 15] 

[[1 2 3]
 [4 5 6]]
[5 7 9]
[ 6 15]


Esto es especialmente útil en análisis por variable (columnas) o por caso (filas).

## Funciones booleanas y de comparación

NumPy permite generar arreglos booleanos que indican si se cumple una condición:

In [18]:
datos = np.array([3, 7, 2, 9])

print(datos > 5)
# [False True False True]
print(np.any(datos > 5))
# True: al menos un valor cumple
print(np.all(datos > 1))
# True: todos los valores > 1
print(np.any(datos <2))
print(np.all(datos > 4))



[False  True False  True]
True
True
False
False


También es posible contar cuántos cumplen la condición:

In [19]:
print(np.sum(datos > 5))
# Resultado: 2
print(np.sum(datos < 10))

2
4


## Funciones universales (ufuncs)

Las funciones matemáticas y estadísticas de NumPy son conocidas como ufuncs o funciones universales. Estas funciones están implementadas en C para operar con alta velocidad sobre cada elemento del arreglo.

Características de las ufuncs:

• Se ejecutan sobre arreglos de cualquier forma o tamaño.

• Admiten operaciones entre arreglos o entre un arreglo y un escalar.

• Se pueden combinar para construir expresiones complejas sin sacrificar velocidad.

In [20]:
x = np.array([1, 2, 3])
y = np.array([4, 5, 6])

print(np.add(x, y)) # sumar vectores
# [5 7 9]
print(np.multiply(x, y)) # multiplicar vectores
# [4 10 18]

[5 7 9]
[ 4 10 18]


Las ufuncs (universal functions) son funciones que:

👉 trabajan elemento por elemento sobre un arreglo NumPy.

Ejemplos típicos de ufuncs:

- np.sqrt

- np.log 

- np.sin

- np.cos

- np.exp

- np.add

- np.maximum

En vez de usar for, NumPy hace esto internamente en C.

In [21]:
datos = np.array([1, 4, 9, 16])

valores = np.sqrt(datos) # calcula la raíz cuadrada.
print(valores)

[1. 2. 3. 4.]


✅ Ahora te explico cada punto de tu imagen

🔹 1. Se ejecutan sobre arreglos de cualquier forma o tamaño

Ejemplo con matriz:

In [22]:
matriz = np.array([[1, 4],
                   [9, 16]])

valor_raiz= np.sqrt(matriz)
print(valor_raiz)


[[1. 2.]
 [3. 4.]]


👉 No importa si es vector o matriz.

🔹 2. Admiten operaciones entre arreglos o entre arreglo y escalar

👉 Arreglo con escalar

In [23]:
datos = np.array([1, 2, 3])
# 👉 Aplica la función exponencial a cada elemento del arreglo.
# np.exp(x) = e^x / 👉 El valor de e es: 2.718281828459045... 👉 e es un número matemático constante como π (pi).
exp = np.exp(datos)
print(exp)


[ 2.71828183  7.3890561  20.08553692]


🎯 En una frase

Este ejercicio toma el arreglo [1,2,3] y calcula:

👉 e¹, e² y e³ usando np.exp, elemento a elemento.

🧠 ¿Qué lo hace especial?

Es el número que aparece de forma natural cuando hay:

crecimiento continuo

procesos que van creciendo poco a poco todo el tiempo

Por ejemplo:

crecimiento de una población

interés compuesto

decaimiento

modelos en ciencia y datos

#### e es un número especial que se usa como base para el crecimiento continuo.

In [24]:
print(datos + 10)
# Esto internamente usa una ufunc (add).


[11 12 13]


👉 Arreglo con arreglo

In [25]:
a = np.array([1, 2, 3])
b = np.array([10, 20, 30])

print(np.add(a, b))


[11 22 33]


👉 Esto es lo mismo que:

In [26]:
print(a + b)


[11 22 33]


### 🔹 3. Se pueden combinar para construir expresiones más complejas

In [27]:
datos = np.array([1, 4, 9])

resultado = np.sqrt(datos) + np.log(datos)
print(resultado)


[1.         3.38629436 5.19722458]


✅ Ejemplos muy parecidos a tus ejercicios

In [28]:
datos = np.array([5, 10, 15])

norm = datos / np.max(datos)
print(norm)


[0.33333333 0.66666667 1.        ]


✔ Seno y coseno (ufuncs reales)

In [29]:
angulos = np.array([0, np.pi/2, np.pi])

seno = np.sin(angulos)
coseno = np.cos(angulos)

print(seno)
print(coseno)


[0.0000000e+00 1.0000000e+00 1.2246468e-16]
[ 1.000000e+00  6.123234e-17 -1.000000e+00]


✅ Ejemplo donde se ve clarísimo el punto 3

In [30]:
x = np.array([1, 2, 3, 4])

y = np.sqrt(x) * np.log(x + 1)
print(y)

[0.69314718 1.5536724  2.40113227 3.21887582]


In [31]:
# esto es para saber que hace cada operación
sq= np.sqrt(x)
lg= np.log(x + 1)
print(sq)
print(lg)

[1.         1.41421356 1.73205081 2.        ]
[0.69314718 1.09861229 1.38629436 1.60943791]


🎯 La idea clave para que te quede

👉 Una ufunc es una función que trabaja:
elemento a elemento sobre arrays.

#### Las ufuncs permiten aplicar funciones matemáticas directamente sobre arreglos, elemento a elemento, sin usar for.

## Composición de funciones

NumPy permite combinar funciones en una sola línea para crear pipelines simples:

En programación y ciencia de datos, el término “pipeline” se refiere a un flujo o secuencia de pasos que se ejecutan de manera organizada y automática, donde la salida de un paso se convierte en la entrada del siguiente. Es como una línea de producción, pero de datos o procesos.

Un pipeline organiza las etapas de transformación de datos y entrenamiento de modelos.

Lectura de datos → Limpieza de datos → Transformación → Entrenamiento del modelo → Predicción

💡 Resumen fácil:
Un pipeline es como una cinta transportadora de pasos automáticos: haces algo en el primer paso, lo que resulta pasa al siguiente, y así hasta terminar todo el proceso.

In [32]:
datos = np.array([1, 4, 9, 16])

resultado = np.log(np.sqrt(datos))
print(resultado)

[0.         0.69314718 1.09861229 1.38629436]


Esto evita la creación de variables intermedias innecesarias y mejora la legibilidad del código.

### Ejemplo: Normalización de datos

Supongamos que recibimos una matriz con mediciones de sensores ambientales y queremos calcular:

• El promedio de cada variable.

• La desviación estándar.

• Un vector con valores normalizados.

In [33]:
''' datos es una matriz de 3×3 (3 filas, 3 columnas).

Cada fila representa una medición de sensores ambientales.

Cada columna representa una variable distinta (por ejemplo, temperatura, humedad, presión).
'''

datos = np.array([
                 [22.5, 55.1, 83.2],
                 [21.9, 53.2, 80.5],
                 [23.0, 54.8, 81.0]
                 ])

media = np.mean(datos, axis=0) # calcula el promedio por columna ej:temperatura, humedad, presión
desviacion = np.std(datos, axis=0) # calcula cuánto se desvían los datos respecto a su media.
normalizado = (datos - media) / desviacion 
# x (datos)= valor original - μ (media)= media de la columna / σ (desviación) = desviación estándar de la columna

print("\nMedia:\n", media)
print("\nDesviacion:\n", desviacion)
print("\nNormalizado:\n", normalizado)



Media:
 [22.46666667 54.36666667 81.56666667]

Desviacion:
 [0.44969125 0.83399973 1.17284081]

Normalizado:
 [[ 0.07412493  0.87929684  1.39263004]
 [-1.26012384 -1.39888134 -0.90947268]
 [ 1.18599891  0.5195845  -0.48315736]]


Cada fila corresponde a la misma fila de datos original, pero ahora los valores son comparables entre columnas, sin importar la escala original.

Por ejemplo:

El valor 22.5 en la primera columna se convirtió en 0.0741

El valor 55.1 en la segunda columna se convirtió en 0.8792

Esto permite que algoritmos de machine learning no se vean sesgados por la magnitud de las variables.

Lo que estás viendo no es un rango 0 a 1, y por eso hay valores mayores a 1 o menores a -1. Esto se debe a que este ejemplo usa normalización Z o estandarización, no normalización min-max. Te explico la diferencia:

### 2️⃣ Media y desviación con axis

np.mean(datos, axis=0) → calcula la media de cada columna.

np.std(datos, axis=0) → calcula la desviación estándar de cada columna.

Esto es lo que usualmente quieres para estandarización, porque cada característica (columna) puede tener unidades y rangos distintos.

- Si no pones axis:

np.mean(datos)  # media de todos los elementos de la matriz
np.std(datos)   # desviación estándar de todos los elementos


Esto mezcla todas las columnas, lo que NO es correcto si quieres escalar cada columna por separado.

💡 Resumen:

axis=0 → escalar por columna (lo usual en ciencia de datos).

axis=1 → escalar por fila (menos común).

Sin axis → calcula sobre toda la matriz, no por características.

### 1️⃣ Normalización Z (o estandarización) ✅

x = valor original

μ = media de la columna

σ = desviación estándar de la columna

Resultado: los valores están centrados en 0 con desviación 1.

Propiedades:

No están limitados entre 0 y 1.

Pueden ser mayores que 1 o menores que -1 si los valores están lejos de la media.

Útil para machine learning porque todas las variables tienen la misma escala relativa.

En tu ejemplo, los valores como 1.3926 o -1.2601 son perfectamente normales con Z-score.

💡 En resumen:

Tu ejemplo no da 0–1 porque usa Z-score, no Min-Max.

Valores mayores a 1 o menores que -1 son normales en la estandarización.

Si necesitas valores estrictamente entre 0 y 1, debes usar Min-Max scaling.

No, normalización y estandarización no son lo mismo, aunque ambos son métodos para escalar datos. Te explico paso a paso con claridad:

### 1️⃣ Normalización (Min-Max Scaling)

Objetivo: Escalar los datos a un rango específico, usualmente [0, 1] o [-1, 1].

Formula :

Xnorm =  X - Xmin / Xmax - Xmin

- Ejemplo:

Datos: [10, 20, 30]

Mínimo = 10, Máximo = 30

Normalizados a [0,1]:

10 → 10 - 10 / 30-10 = 0/20 = 0

20 → 20 - 10 / 30-10 = 10/20 = 0.5

30 → 30-10 / 30-10 = 20/20 = 1

Uso típico: cuando quieres que todos los valores estén dentro de un rango fijo, útil para algoritmos como redes neuronales o KNN.


### 2️⃣ Estandarización (Z-score Scaling)

Objetivo: Transformar los datos para que tengan media 0 y desviación estándar 1.

Fórmula:

Xest = X-μ / σ

donde 
𝜇 = media, 

σ = desviación estándar.

- Ejemplo:

Datos: [10, 20, 30]

Media = 20, Desviación estándar ≈ 8.16

- Estandarizados:

10 → (10-20)/8.16 ≈ -1.22

20 → (20-20)/8.16 = 0

30 → (30-20)/8.16 ≈ 1.22

Uso típico: cuando los algoritmos asumen distribución normal de los datos (como regresión lineal, SVM, PCA).

| Método          | Rango             | Media/Desviación                   | Uso típico                       |
| --------------- | ----------------- | ---------------------------------- | -------------------------------- |
| Normalización   | 0 a 1 (o -1 a 1)  | No cambia la forma de distribución | Redes neuronales, KNN            |
| Estandarización | No hay rango fijo | Media=0, Desviación=1              | Algoritmos que asumen normalidad |
